### ISSUE TITLE
Tried to load big dataset, napari uses all CPU/RAM, forces hard shutdown

### ISSUE THREAD

**Issue by aphilip442** [napari/napari#2489]
## 🐛 Bug

I tried to load a collection of tiff files, amounting to 1TB in total.
Napari proceeded to fill the RAM (256GB), then bottlenecked the CPU as well.
Windows became totally unresponsive (impossible to do anything), forcing a hard shutdown.

## To Reproduce

1. Open Napari
2. Click 'Open Files as Stack'
3. Select amount of files that exceed your RAM (in my case 1TB data for 256GB RAM)
4. See Napari fill the RAM in task manager, then bottleneck the CPU as well
5. Windows is now unresponsive and unrecoverable, you have to shut it down manually by pressing the power button

## Expected behavior

I expected Napari to check for the available RAM and throwing an error/warning message, instead of crashing the computer.
Optionally, I would have preferred for the program to crash.

If this behaviour is considered normal (maybe the sample was actually loading and would have been viewable in XXX time?), there should be an indication in the GUI, the GUI should remain responsive, and the machine's resources should not be all bottlenecked at 100%.

## Environment

 - Please copy and paste the information at napari info option in help menubar here:

napari: 0.4.7
Platform: Windows-10-10.0.14393-SP0
Python: 3.9.2 (default, Mar 3 2021, 15:03:14) [MSC v.1916 64 bit (AMD64)]
Qt: 5.12.9
PyQt5: 5.12.3
NumPy: 1.20.2
SciPy: 1.6.2
Dask: 2021.03.1
VisPy: 0.6.6

OpenGL:
- GL version: 4.6.0 NVIDIA 461.72
- MAX_TEXTURE_SIZE: 32768

Screens:
- screen 1: resolution 1920x1080, scale 1.0

Plugins:
- console: 0.0.3
- svg: 0.1.4

**Comment by jni** [comment]
@aphilip442 could you tell us the dimensions of the individual tiff files + the full stack? I agree that this shouldn't have happened if a single view is smaller than RAM, which I suspect it is.

**Comment by jni** [comment]
(I *also* agree that we could be smarter about dataset size + RAM, but we've often relied on Python to handle that for us. Sometimes it doesn't do so well!)

**Comment by sofroniewn** [comment]
Note for debugging purposes this is equivalent to calling `viewer.open(path, stack=True)` where `path` is the path.

https://github.com/napari/napari/blob/fee29a32b05499466cd2f6e9192cb0c43c92011f/napari/components/viewer_model.py#L661

It doesn't look like there are any extra plugins installed, so this will just use our builtin (but that could have been a source of error). That should lead to a call to `magic_imread` with a list of files and its default arguments. https://github.com/napari/napari/blob/102a7e8f845893c874d2b86f9371d41130100b89/napari/utils/io.py#L141

That should trigger the usage of dask here 

https://github.com/napari/napari/blob/102a7e8f845893c874d2b86f9371d41130100b89/napari/utils/io.py#L212-L214

and prevent the CPU/ RAM explosion, but I guess something is going wrong ....

It would be good to figure out if that line is indeed getting called, or if something else bad is happening before then. I feel reasonably confident we should be able to fix this. Thanks for the report @aphilip442!!

**Comment by aphilip442** [comment]
@jni individual files are 350MB tiff, totaling 1TB for the full stack. 
I was testing how napari would handle a large dataset, and was indeed expecting lazy loading, since the data far exceeds the available RAM. 
@sofroniewn happy to provide the report!

**Comment by jni** [comment]
@aphilip442 what is the shape of the individual files, and was the viewer in 2D view or 3D view?

@sofroniewn I've experienced a few situations, loading zarr, where napari incorrectly loads stacks as layers, and I worry that is what is happening here.

We should also change that `da.stack` to a `da.map_blocks`.

**Comment by aphilip442** [comment]
@jni the files are 16626x11084pixels, 16-bit, monochrome. The viewer was in 2D mode.
Additional note: I switched the mode back and forth (if that could have any incidence), the sequence was like this : launched napari -> loaded individual file -> 3D -> 2D -> remove small file -> tried to load large dataset -> then bug report.

**Comment by sofroniewn** [comment]
> 16626x11084pixels

Looks like the individual 2D files were quite big too!! We should be able to handle them, but this will be much better with our async/ tiled viewing enabled https://napari.org/guides/stable/rendering.html#rendering. (note I still think there is an underlying bug here, and that is more of an aside)

**Comment by jni** [comment]
@aphilip442 Q: if you make a smaller folder with just a few tiffs, do they load as a stack or as layers?

**Comment by aphilip442** [comment]
@jni A: there is only one layer in the layer list, with scrollable frames in the view (with 'play' buttons). I suppose that would be a stack.

**Comment by aphilip442** [comment]
Additional note: in Ubuntu 20.04 the files are loaded lazily without issue

**Comment by jni** [comment]
@aphilip442 that last tidbit is very weird indeed...! Are you sure there's no difference in dask/napari/tifffile versions in the Linux vs Windows environment?

**Comment by sofroniewn** [comment]
@jni I wonder now that we've turned off the dask cache - see #2590 if this might be better? Apologies if that's already been tried

**Comment by aphilip442** [comment]
@jni here are the environment files, both installation were done on fresh installs of miniconda, using the 'conda install napari -c conda-forge' command, without warnings/errors. Nothing else was installed in those environments.

[ubuntu_env.txt](https://github.com/napari/napari/files/6364161/ubuntu_env.txt)
[windows_env.txt](https://github.com/napari/napari/files/6364162/windows_env.txt)

**Comment by kephale** [comment]
I think this one needs to be re-verified because a lot of the relevant codebase has changed. @ClementCaporal has some tiffs for testing and is looking into it!

**Comment by ClementCaporal** [comment]
Hello!
TLDR; I cannot reproduce the error

I tried on a `Windows 11, version 23H2`. Fresh install using `mamba install napari[all] -c conda-forge`, python 3.10
I opened a folder as stack of 214 files of individual size 450MB (74, 3, 1024, 1024) (z, c, y, x) for a total 91.1GB using:
- `File>Open Files as Stack`
- drag and drop the folder

It used very few of the RAM (less than 3GB) so I suppose it delays it correctly.
